runtime/runtime/src/lib.rs

```diff
@@ -155,6 +155,79 @@ pub struct ApplyState {
     pub bandwidth_requests: BlockBandwidthRequests,
 }
 
+/// Represents a batch of transactions sharing the same (signer_id, public_key),
+/// sorted by ascending nonce.
+struct TransactionBatch<'a> {
+    pub indices: &'a [usize],
+    pub signed_txs: &'a [SignedTransaction],
+}
+
+impl<'a> TransactionBatch<'a> {
+    fn iter(&self) -> impl Iterator<Item = (usize, &SignedTransaction)> {
+        self.indices.iter().map(|idx| (*idx, &self.signed_txs[*idx]))
+    }
+}
+
+/// Represents a collection of transaction batches. Provides a parallel iterator.
+struct TransactionBatches<'a> {
+    signed_txs: &'a [SignedTransaction],
+    indices: Vec<usize>,
+}
+
+impl<'a> TransactionBatches<'a> {
+    /// Constructs TransactionBatches from a slice of SignedTransaction.
+    /// Transactions are sorted by (signer_id, public_key, nonce).
+    pub fn new(signed_txs: &'a [SignedTransaction]) -> Self {
+        let mut indices: Vec<usize> = (0..signed_txs.len()).collect();
+        indices.sort_by_key(|&i| {
+            let tx = &signed_txs[i].transaction;
+            (tx.signer_id(), tx.public_key(), tx.nonce())
+        });
+        Self { signed_txs, indices }
+    }
+
+    /// Returns a parallel iterator over transaction batches batched by (signer_id, public_key).
+    pub fn par_batches(
+        &'a self,
+    ) -> impl rayon::iter::ParallelIterator<Item = TransactionBatch<'a>> {
+        self.indices
+            .par_chunk_by(|&left, &right| {
+                let left_tx = &self.signed_txs[left].transaction;
+                let right_tx = &self.signed_txs[right].transaction;
+                left_tx.signer_id() == right_tx.signer_id()
+                    && left_tx.public_key() == right_tx.public_key()
+            })
+            .map(|chunk| TransactionBatch { indices: chunk, signed_txs: self.signed_txs })
+    }
+}
+
+#[derive(Debug)]
+struct ProcessedTransaction {
+    index: usize,
+    transaction: ValidatedTransaction,
+    verification_result: VerificationResult,
+    receipt: Receipt,
+    outcome: ExecutionOutcomeWithId,
+}
+
+#[derive(Debug)]
+struct BatchResult {
+    processed_transactions: Vec<ProcessedTransaction>,
+    final_state: Option<(Account, AccessKey)>,
+}
+
+impl BatchResult {
+    /// Returns the signer_id of the first processed transaction in this batch, if any.
+    pub fn signer_id(&self) -> Option<&AccountId> {
+        self.processed_transactions.first().map(|pt| pt.transaction.to_tx().signer_id())
+    }
+
+    /// Returns the public_key of the first processed transaction in this batch, if any.
+    pub fn public_key(&self) -> Option<&PublicKey> {
+        self.processed_transactions.first().map(|pt| pt.transaction.to_tx().public_key())
+    }
+}
```

Comment:
these are used only once in production code, I would argue that we can get rid of them.

Reply:
I think these methods do a great job of abstracting away icky logic that's entirely an implementation detail of `BatchResult`. Business code has no reason to know about details of how batching is implemented.